docs/style-guide.rst

```diff
@@ -51,90 +50,558 @@ into the making of Hy.
+               "not a number"))
+
+    ;; Worse.
+    (print (if (< n 0.0) "negative"
+               (= n 0.0) "zero"
+               (and (even? n)
+                    (> n 0.0)) (do (do-foo)  ; Group not started this line.
+                                   (do-bar)
+                                   "even-positive")
+               (> n 0.0) "positive"
+               "not a number"))
+
+    ;; Good. Blank line separates groups.
+    (print (if (< n 0.0) "negative"
+
+               (= n 0.0) "zero"
+
+               (and (even? n)
+                    (> n 0.0))
+               (do (do-foo)
+                   (do-bar)
+                    "even-positive")
+
+               (> n 0.0) "positive"
+
+               "not a number"))
+
+    ;; Not so good, groups are not separated consistently.
+    (print (if (< n 0.0) "negative"
+               (= n 0.0) "zero"
+
+               (> n 0.0)
+               (do (do-foo)
+                   "positive")
+
+               "not a number"))
+
+    ;; Acceptable. All groups are separated the same way, with a blank like.
+    (print (if (< n 0.0) "negative"
+
+               (= n 0.0) "zero"
+
+               (> n 0.0)
+               (do (do-foo)
+                   "positive")
+
+               "not a number"))
+
+    (defn fib [n]  ; saw a "]", newline.
+      (if (<= n 2)  ; OK to break here. Since there's only one pair, we don't have to separate them.
+          n
+        (+ (fib (- n 1))  ; non-special indent is another whitespace separation technique.
+           (fib (- n 2)))))
 
-    ;; Hysterically ridiculous
-    (defn fib [n]
-      (if (<= n 2)
-        n
-        (+ (fib (- n 1)) (fib (- n 2)))
-      )
-    )  ; GAH, BURN IT WITH FIRE
+Comments
+--------
 
+Prefer docstrings to comments where applicable--in ``defn``, ``defclass``, and at the top of the module.
 
-+ Inline comments shall be two spaces from the end of the code; they
-  must always have a space between the comment character and the start
-  of the comment. Also, try to not comment the obvious.
+The ``(comment)`` macro is still subject to the three laws.
+If you're tempted to violate them, consider discarding a string instead with ``#_``.
+
+Semicolon comments shall start with some number of semicolons
+and have a space between the semicolons and the start of the comment.
+Also, try to not comment the obvious.
 
 .. code-block:: clj
 
-   ;; Good
-   (setv ind (dec x))  ; indexing starts from 0
+    ;;;; Major Header Labeling a Major Section
+    ;; Headers should only be one line.
+    ;; This is non-header commentary, but not about a particular form.
+    ;; These can span multiple lines.
+    ;; These are separated from the next form or form comment by a blank line.
 
-   ;; Style-compliant but just states the obvious
-   (setv ind (dec x))  ; sets index to x-1
+    ;; Good.
+    (setv ind (dec x))  ; indexing starts from 0
+                                            ; margin comment continues on the next line.
```

Comment:
Was the extra indentation accidental or intentional? I don't see anything here really justifying that.

Reply:
Yeah we should probably discuss that a bit more. The Clojure Style Guide (linked in the thanks section at the bottom, even in the current version) uses the term "[margin comments](https://github.com/bbatsov/clojure-style-guide#one-semicolon-for-margin-comments)".

@ekaschalk Emacs likes to indent Hy's `;` comments to column 32, but it's column 40 in clojure-mode, and lisp-mode. I'm pretty sure I've seen column-aligned comments like this in Common Lisp code. I think they call them "column comments". How would you want to handle these?

Should "margin comments" be a separate idea from the "column comments"? Should we forbid continuing margin comments on an otherwise blank line in the style guide? Should we get rid of the 2-space margin comments and only recommend column comments? Something else?